pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/ColocatedRealtimeSegmentAssignment.java

```diff
@@ -0,0 +1,123 @@
+import java.util.Map;
+import java.util.TreeMap;
+import javax.annotation.Nullable;
+import org.apache.commons.configuration.Configuration;
+import org.apache.helix.HelixManager;
+import org.apache.pinot.common.assignment.InstancePartitions;
+import org.apache.pinot.common.tier.Tier;
+import org.apache.pinot.common.utils.SegmentUtils;
+import org.apache.pinot.spi.config.table.ReplicaGroupStrategyConfig;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.assignment.InstancePartitionsType;
+import org.apache.pinot.spi.utils.CommonConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class ColocatedRealtimeSegmentAssignment implements SegmentAssignment {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ColocatedRealtimeSegmentAssignment.class);
+
+  private HelixManager _helixManager;
+  private String _realtimeTableName;
+  private int _replication;
+  private String _partitionColumn;
+
+  @Override
+  public void init(HelixManager helixManager, TableConfig tableConfig) {
+    _helixManager = helixManager;
+    _realtimeTableName = tableConfig.getTableName();
+    _replication = tableConfig.getValidationConfig().getReplicasPerPartitionNumber();
+    ReplicaGroupStrategyConfig replicaGroupStrategyConfig =
+        tableConfig.getValidationConfig().getReplicaGroupStrategyConfig();
+    _partitionColumn = replicaGroupStrategyConfig != null ? replicaGroupStrategyConfig.getPartitionColumn() : null;
+
+    LOGGER.info("Initialized ColocatedRealtimeSegmentAssignment with replication: {}, partitionColumn: {} for table: "
+            + "{}",
+        _replication, _partitionColumn, _realtimeTableName);
+  }
+
+  @Override
+  public List<String> assignSegment(String segmentName, Map<String, Map<String, String>> currentAssignment,
+      Map<InstancePartitionsType, InstancePartitions> instancePartitionsMap) {
+    InstancePartitions instancePartitions = instancePartitionsMap.get(InstancePartitionsType.CONSUMING);
+    Preconditions.checkState(instancePartitions != null, "Failed to find CONSUMING instance partitions for table: %s",
+        _realtimeTableName);
+    Preconditions
+        .checkState(instancePartitions.getNumPartitions() == 1, "Instance partitions: %s should contain 1 partition",
+            instancePartitions.getInstancePartitionsName());
+    LOGGER.info("Assigning segment: {} with instance partitions: {} for table: {}", segmentName, instancePartitions,
+        _realtimeTableName);
+    checkReplication(instancePartitions);
+
+    int partitionGroupId =
+        SegmentUtils.getRealtimeSegmentPartitionId(segmentName, _realtimeTableName, _helixManager, _partitionColumn);
+    List<String> instances = SegmentAssignmentUtils.assignSegmentForCoLocatedTable(instancePartitions,
+        partitionGroupId);
+    return instances;
+  }
+
+  @Override
+  public Map<String, Map<String, String>> rebalanceTable(Map<String, Map<String, String>> currentAssignment,
```

Comment:
could you explain how we achieve the cross-table coordination?

Reply:
Right now there's no cross-table coordination.

The rebalance flow is as follows:

1. Say users adds new instances to a tenant. They first need to go to the Table Group UI page, and click on Recompute Table Group. This will update the instances assigned for that table-group.
2. Next, they'll need to Rebalance each of the tables individually. This is where this function is called.

So essentially right now this function works on a per-table basis.

I haven't yet started thinking about cross-table coordination (or a group rebalance)... will do it this week.